### ISSUE TITLE
Training hanged at loss.backward() of movie-dialog-qa

### ISSUE THREAD

**Issue by jefst** [facebookresearch/ParlAI#307]
Hi, I am interested in Memory Networks and Movie Dialog QA. Although I tried below command, training stopped at first `loss.backward()` at `memnn.py`. I waited almost one day, but `loss.backward()` didn't finish. I have checked this by printing debug and `[Using Cuda]` printing. Actually my GPU was working because it used some memory. I checked this by `nvidia-smi -l 1`.

`python examples/train_model.py -m memnn -t "#moviedd-qa" -bs 32 --gpu 0 -e 10`

Then, I switched to simple task, and it finished a few minutes.
`python examples/train_model.py -m memnn -t "babi:task1k:1" -bs 32 --gpu 0 -e 10 `

I realize `#moviedd-qa` is more complicated compared to babi task. But how long does it usually take to train this model in my setting? Does anyone try to train this model via ParlAI? I am afraid this is not bug of ParlAI. Could you advise me to proceed my work?

### My Environment
- Ubunt 16.04.03 LTS, 64 bit
- python 3.6.1 (Anaconda 4.4.0 (64-bit))
- GPU: GTX 1080 ti
- CPU: Intel(R) Core(TM) i7-7700K CPU @ 4.20GHz
- torch.\_\_version\_\_: '0.2.0_3'

**Comment by alexholdenmiller** [comment]
Hi @jefst, thanks for flagging! We'll take a look when we get the chance. We've tested in quite a bit on babi, but we hadn't yet tried to reproduce the moviedd-qa results with this new pytorch implementation of the memory network.

**Comment by jojonki** [comment]
Hi, thank you for your reply. I hope there will be some progress on this task, including #313 . 
So far, I am struggling to reproduce this task. 
[Evaluating Prerequisite Qualities for Learning End-to-End Dialog Systems](https://arxiv.org/abs/1511.06931)

@jefst is exactly me. I posted at my wrong account. I am sorry for the confusing.

**Comment by alexholdenmiller** [comment]
@jojonki to be clear, we found that this was just due to the size of the candidate set. For this task, there were ~75k candidates, so it took the memnn a long time to rank them all--on my GPU, I think it took around an hour to do a batch of 32...

**Comment by alexholdenmiller** [comment]
So now we just sample a small set of candidates to use for training.

**Comment by jojonki** [comment]
@alexholdenmiller 
Thank you for your help! The training speed is now much faster! But the performance was really low. Is this the problem of the number of epoch? I doubt there are other wrong things.

```
$ python examples/train_model.py -m memnn -t "#moviedd-qa" -bs 32 --gpu 0 -e 10
[ Main ParlAI Arguments: ] 
[  task: #moviedd-qa ]
[  download_path: /home/jonki/work/ParlAI/downloads ]
[  datatype: train ]
[  image_mode: raw ]
[  numthreads: 1 ]
[  batchsize: 32 ]
[  datapath: /home/jonki/work/ParlAI/data ]
[ ParlAI Model Arguments: ] 
[  model: memnn ]
[  model_file: None ]
[  dict_class: None ]
[ Dictionary Arguments: ] 
[  dict_file: None ]
[  dict_initpath: None ]
[  dict_language: english ]
[  dict_max_ngram_size: -1 ]
[  dict_minfreq: 0 ]
[  dict_nulltoken: __NULL__ ]
[  dict_starttoken: __START__ ]
[  dict_endtoken: __END__ ]
[  dict_unktoken: __UNK__ ]
[  dict_maxexs: 100000 ]
[ MemNN Arguments: ] 
[  learning_rate: 0.01 ]
[  embedding_size: 128 ]
[  hops: 3 ]
[  mem_size: 100 ]
[  time_features: True ]
[  position_encoding: False ]
[  output: rank ]
[  rnn_layers: 2 ]
[  dropout: 0.1 ]
[  optimizer: adam ]
[  no_cuda: False ]
[  gpu: 0 ]
[ Training Loop Arguments: ] 
[  evaltask: None ]
[  display_examples: False ]
[  num_epochs: 10.0 ]
[  max_train_time: -1 ]
[  log_every_n_secs: 2 ]
[  validation_every_n_secs: -1 ]
[  validation_max_exs: -1 ]
[  validation_patience: 10 ]
[  validation_metric: accuracy ]
[  dict_build_first: True ]
[ building dictionary first... ]
Tried to build dictionary but `--dict-file` is not set. Set this param so the dictionary can be saved.
[ Using CUDA ]
[creating task(s): moviedialog:Task:1]
[loading fbdialog data:/home/jonki/work/ParlAI/data/MovieDialog/movie_dialog_dataset/task1_qa/task1_qa_pipe_train.txt]
[ Using CUDA ]
...
[ time:2s parleys:5 total_exs:160 time_left:13944s ] {'total': 160, 'accuracy': 0.01875, 'f1': 0.01875, 'hits@k': {1: 0.0187, 5: 0.0
75, 10: 0.156, 100: 1.0}}
[ time:4s parleys:12 total_exs:384 time_left:11322s ] {'total': 224, 'accuracy': 0.008929, 'f1': 0.008929, 'hits@k': {1: 0.00893, 5:
 0.067, 10: 0.143, 100: 1.0}}
[ time:6s parleys:19 total_exs:608 time_left:10679s ] {'total': 224, 'accuracy': 0.004464, 'f1': 0.007228, 'hits@k': {1: 0.00446, 5:
 0.0446, 10: 0.121, 100: 1.0}}
...
...
[ time:9415s parleys:30046 total_exs:961472 time_left:3s ] {'total': 224, 'accuracy': 0.004464, 'f1': 0.004464, 'hits@k': {1: 0.00446, 5: 0.0536, 10: 0.152, 100: 1.0}}
[ time:9417s parleys:30053 total_exs:961696 time_left:1s ] {'total': 224, 'accuracy': 0.03571, 'f1': 0.03795, 'hits@k': {1: 0.0357, 5: 0.0893, 10: 0.183, 100: 1.0}}
[ num_epochs completed: 10.0 ]
[ running eval: valid ]
[creating task(s): moviedialog:Task:1]
[loading fbdialog data:/home/jonki/work/ParlAI/data/MovieDialog/movie_dialog_dataset/task1_qa/task1_qa_pipe_dev.txt]
[ Using CUDA ]
...
valid:{'total': 9968, 'accuracy': 0.0001003, 'f1': 0.00222, 'hits@k': {1: 0.0001, 5: 0.000502, 10: 0.000602, 100: 0.0015}}
[ running eval: test ]
[creating task(s): moviedialog:Task:1]
[loading fbdialog data:/home/jonki/work/ParlAI/data/MovieDialog/movie_dialog_dataset/task1_qa/task1_qa_pipe_test.txt]
[ Using CUDA ]
...
test:{'total': 9952, 'accuracy': 0.0001005, 'f1': 0.002589, 'hits@k': {1: 0.0001, 5: 0.0001, 10: 0.000402, 100: 0.00171}}
```
